File: 146-lru-cache/lru-cache.py
class LRUCache:

    class Node:
        def __init__(self, key, value, back, next) -> None:
            self.key = key
            self.value = value
            self.back = back
            self.next = next

    def __init__(self, capacity: int):
        self.header = self.Node(-1, -1, None, None)
        self.footer = self.Node(-1, -1, None, None)
        self.capacity = capacity
        self.hash = dict()
        self.size = 0
        

    def get(self, key: int) -> int:
        if key in self.hash:
            self.put(key, self.hash[key].value)
            return self.hash[key].value
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.hash:
            node = self.hash[key]
            # remove node from ll
            next = node.next
            back = node.back
            back.next = next
            next.back = back
            self.size -= 1
        elif self.size == self.capacity:
            to_delete = self.footer.next
            self.hash.pop(to_delete.key)
            next = to_delete.next
            self.footer.next = next
            next.back = self.footer
            self.size -= 1

        new_node = self.Node(key, value, None, None)
        self.hash[key] = new_node
        
        if self.size == 0:
            self.header.back =  new_node
            self.footer.next = new_node
            new_node.next = self.header
            new_node.back = self.footer
        else:
            prev = self.header.back
            prev.next = new_node
            new_node.next = self.header
            self.header.back = new_node
            new_node.back = prev
        self.size += 1

File: 146-lru-cache/lru-cache.cpp
struct dll {
    dll* prev;
    dll* next;
    int key;
    int value;
};

class LRUCache {
public:
    int capacity; 
    dll* head;
    dll* tail;
    int occupancy;
    unordered_map<int, dll*> lru_map;

    LRUCache(int capacity) {
        this->capacity = capacity;
        occupancy = 0;
        head = nullptr;
        tail = nullptr;
        lru_map = unordered_map<int, dll*>();
    }

    void pushLeft(int key, int value){
        dll* node = new dll(nullptr, nullptr, key, value);
        pushLeft(node);
    }

    void pushLeft(dll* node){
        lru_map[node->key] = node;
        if(head != nullptr) {
            node->next = head;
            head->prev = node;
            head = node;
        } else{
            head = node;
            tail = node;
        }
        occupancy += 1;
    }

    dll* popRight() {
        dll* node = tail;

        if (node == nullptr) return nullptr;
        lru_map.erase(node->key);
        occupancy -= 1;

        dll* prev = node->prev;
        if(prev == nullptr){
            head = nullptr;
            tail = nullptr;
        } else {
            prev->next = nullptr;
            tail = prev;
        }
        return node;
    }

    dll* popNode(dll* node){
        if(node == nullptr || !lru_map.contains(node->key)) return nullptr;

        if(node == tail){
            return popRight();
        } 
        lru_map.erase(node->key);
        occupancy -= 1;
        if(node == head) {
            dll* next = node->next;
            next->prev = nullptr;
            head = next;
            return node;
        }
        dll* prev = node->prev;
        dll* next = node->next;
        prev->next = next;
        next->prev = prev;
        return node;
    }

    void promote(dll* node){
        popNode(node);
        pushLeft(node);
    }
    
    int get(int key) {
        if(lru_map.contains(key)){
            promote(lru_map[key]);
            return lru_map[key]->value;
        }
        return -1;
    }

    void put(int key, int value) {
        if(lru_map.contains(key)){
            lru_map[key]->value = value;
            promote(lru_map[key]);
            return;
        }
        if(occupancy == capacity){
            dll* poped_node = popRight();
            delete(poped_node);
            pushLeft(key, value);
            return;
        }
        pushLeft(key, value);
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */